gtk/Session.cc

```diff
@@ -1222,29 +1224,60 @@ void Session::Impl::send_rpc_request(
 
 void Session::port_test()
 {
-    auto const tag = nextTag;
-    ++nextTag;
+    enum : uint8_t
+    {
+        // These numbers are used as array indices
+        PORT_TEST_IPV4,
+        PORT_TEST_IPV6,
+        NUM_PORT_TEST_IP_PROTOCOL
+    };
 
-    tr_variant request;
-    tr_variantInitDict(&request, 2);
-    tr_variantDictAddStrView(&request, TR_KEY_method, "port-test");
-    tr_variantDictAddInt(&request, TR_KEY_tag, tag);
-    impl_->send_rpc_request(
-        request,
-        tag,
-        [this](auto& response)
-        {
-            tr_variant* args = nullptr;
-            bool is_open = false;
+    static auto constexpr IpStr = std::array{ "ipv4"sv, "ipv6"sv };
+    static auto tags = std::array<std::optional<int64_t>, NUM_PORT_TEST_IP_PROTOCOL>{};
+    static auto results = std::array<std::optional<bool>, NUM_PORT_TEST_IP_PROTOCOL>{};
```

Comment:
`tags` is there so that even if somebody clicked on the port test button again before the previous `port-test` requests finished (which is possible by immediately re-opening the pref dialogue after clicking the button for the first time), the status label text will only update in response to the newest `port-test` RPC calls.

Not sure how I can do that by relying on the existing infrastructure.

Reply:
Theoretically, can have some kind of flag in `Session` that says that port test request is in flight (either a separate bool or the same shared_ptr stored as a member and reset on request finish), and then either
* ignore new requests while one is active; this would mean clicking the button second time will disable it but no actual request will be sent, and callback from active request will enable it back
* expose the flag and check for it on `PrefsDialog` init to disable the button if needed

Reply:
012e7e0 GTK
099da59 Qt